content/typeclasses-syntax.md

```diff
@@ -0,0 +1,657 @@
+
+**Benefits:**
+
+ - Better orthogonality, type parameters and abstract type members now accept the same kinds of bounds.
+ - Better ergonomics, since deferred givens get naturally implemented in inheriting classes, no need for boilerplate to fill in definitions of abstract givens.
+
+**Alternative:** It was suggested that we use a modifier for a deferred given instead of a `= deferred`. Something like `deferred given C[T]`. But a modifier does not suggest the concept that a deferred given will be implemented automatically in subclasses unless an explicit definition is written. In a sense, we can see `= deferred` as the invocation of a magic macro that is provided by the compiler. So from a user's point of view a given with `deferred` right hand side is not abstract.
+It is a concrete definition where the compiler will provide the correct implementation.
+
+### 5. Context Bounds for Polymorphic Functions
+
+Currently, context bounds can be used in methods, but not in function types or function literals. It would be nice  propose to drop this irregularity and allow context bounds also in these places. Example:
+
+```scala
+type Comparer = [X: Ord] => (x: X, y: X) => Boolean
+val less: Comparer = [X: Ord as ord] => (x: X, y: X) =>
+  ord.compare(x, y) < 0
+```
+
+The expansion of such context bounds is analogous to the expansion in method types, except that instead of adding a using clause in a method, we insert a context function type.
+
+For instance, type and val above would expand to
+```scala
+type Comparer = [X] => (x: X, y: X) => Ord[X] ?=> Boolean
+val less: Comparer = [X] => (x: X, y: X) => (ord: Ord[X]) ?=>
+  ord.compare(x, y) < 0
+```
+
+The expansion of using clauses does look inside alias types. For instance,
+here is a variation of the previous example that uses a parameterized type alias:
+```scala
+type Cmp[X] = (x: X, y: X) => Ord[X] ?=> Boolean
+type Comparer2 = [X: Ord] => Cmp[X]
+```
+The expansion of the right hand side of `Comparer2` expands the `Cmp[X]` alias
+and then inserts the context function at the same place as what's done for `Comparer`.
+
+### 6. Cleanup of Given Syntax
+
+A good language syntax is like a Bach fugue: A small set of motifs is combined in a multitude of harmonic ways. Dissonances and irregularities should be avoided.
+
+When designing Scala 3, I believe that, by and large, we achieved that goal, except in one area, which is the syntax of givens. There _are_ some glaring dissonances, as seen in this code for defining an ordering on lists:
+```scala
+given [A](using Ord[A]): Ord[List[A]] with
+  def compare(x: List[A], y: List[A]) = ...
+```
+The `:` feels utterly foreign in this position. It's definitely not a type ascription, so what is its role? Just as bad is the trailing `with`. Everywhere else we use braces or trailing `:` to start a scope of nested definitions, so the need of `with` sticks out like a sore thumb.
```

Comment:
The `:` feels utterly familiar in this position to me.

```scala
def foo()[A](using Ord[A]): Ord[List[A]]
```

What is foreign about that?  We write this ubiquitously!  It's simply the return type.

What is weird is that we're both declaring the return type and defining an anonymous class that implements it and asking for an instance to be created _all with the same `: Type with` construct.  That _is_ a lot of lifting for not very much syntax.

The non-given way we'd do that is by leveraging type inference

```scala
def foo[A](using Ord[A]) = new Ord[List[A]]:
  ...
```

Reply:
```scala
given [A](using Ord[A]): Ord[List[A]] with
```
the problem is that it isn't a return type, its an extends clause in disguise (here is the desugar):
```scala
implicit def Ord_List[A](using Ord[A]): Ord_List[A] = new Ord_List[A]

class Ord_List[A](using Ord[A]) extends Ord[List[A]]:
  def compare(x: List[A], y: List[A]) =
```

Reply:
Okay, but do we actually _want_ this to be a `given_Ord_List[A]` as opposed to an `Ord[List[A]]`?  If I were aliasing, I wouldn't have declared a new given-specific type.  This is pretty weird, because I can

```scala
var x = summon[Ord[List[Int]]]
```

and depending on whether I used an alias or created it de-novo, I get a different return type.

That's kind of...baffling.  You can _still_ explain the `given ... with` as doing a `transparent` as well, if you want to say that `: Ord[List[A]]` is doing its normal job (I'd still say it is).  So that _still_ doesn't seem to be the weird part to me.

Anyway, I don't understand why one would want to expose in this fashion what looks like implementation details.

Reply:
because the with can introduce new refinements

Reply:
The question still stands as to whether you want to expose this in a given.  Because you can't easily name the type, you can't declare that you intend to get those refinements.  You only get them accidentally when you ask for the named type from a context where the refinement has visibility.

Depending on the details of behavior of an un-nameable type seems like an antipattern.

Note that this is not the case for vals in Scala 3.4.2:

```scala
scala> trait T { def foo: Int }
// defined trait T
                                                                                
scala> val x = new T:
     |   val foo = 5
     |   def bar = true
     | 
val x: T = anon$1@52021702
                                                                                
scala> x.bar
-- [E008] Not Found Error: -----------------------------------------------------
1 |x.bar
  |^^^^^
  |value bar is not a member of T
1 error found
```

Not true for defs:

```scala
     |   val foo = i
     |   def bar = i < 0
     | 
def f(i: Int): T
                                                                                
scala> f(3).bar
-- [E008] Not Found Error: -----------------------------------------------------
1 |f(3).bar
  |^^^^^^^^
  |value bar is not a member of T
1 error found
```

So that the refinement is visible is an inconsistency that perhaps should be removed.  Do we crucially depend upon it for anything?

Reply:
actually this is identical to implicit object in scala 2 - and the name is visible in scope even for anonymous, or just don't make it anonymous?

Reply:
That it parallels Scala 2 is less important, I think, than being consistent with Scala 3.  (Note that there was a behavior change for defs and vals; you used to get a return type that was the supertype with a structural type over the refinements.)